lib/common/output_text.c

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2019 Chris Lumens <clumens@redhat.com>
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <stdlib.h>
+#include <crm/crm.h>
+#include <crm/common/output.h>
+
+typedef struct text_list_data_s {
+    unsigned int len;
+    char *singular_noun;
+    char *plural_noun;
+} text_list_data_t;
+
+typedef struct text_private_s {
+    GQueue *parent_q;
+} text_private_t;
+
+static void
+indent(FILE *dest, int level) {
+    for (int i = 0; i < level; i++) {
+        fprintf(dest, "\t");
+    }
+}
+
+static void
+pcmk__free_text_priv(void *out) {
+    text_private_t *priv = (text_private_t *) out;
+
+    if (priv == NULL) {
+        return;
+    }
+
+    g_queue_free(priv->parent_q);
+    free(priv);
+}
+
+static gboolean
+text_init(pcmk__output_fmt_t *out) {
+    text_private_t *priv = NULL;
+
+    /* If text_init was previously called on this output struct, just return. */
+    if (out->priv != NULL) {
+        return true;
+    } else {
+        out->priv = calloc(1, sizeof(text_private_t));
+        if (out->priv == NULL) {
+            return false;
+        }
+
+        priv = out->priv;
+    }
+
+    priv->parent_q = g_queue_new();
+    return true;
+}
+
+static void
+text_finish(pcmk__output_fmt_t *out, crm_exit_t exit_status) {
+    /* This function intentionally left blank */
+}
+
+static void
+text_reset(pcmk__output_fmt_t *out) {
+    CRM_ASSERT(out->priv != NULL);
+
+    pcmk__free_text_priv(out->priv);
+    text_init(out);
```

Comment:
Looking at this, I'm wondering how reset should behave.

The main use will be for crm_mon's curses output. reset() would clear the screen and start from the top again. This implies that we always want to reset at the beginning of a round of output, rather than the end (because the screen would be immediately wiped in that case). For example, a typical flow would be:

    out = new()
    while (keep_going) {
        reset(out)
        some_output(out)
    }
    free(out)

At this point, I don't think it matters what other formats do -- if someone has a particular need, we can revisit. But theorizing about it, for stdout (or a named pipe), it's pretty obvious that we'd just want to print any ending info and keep going, without touching the file descriptor. For a file, I could see two possibilities: the same (i.e. the file will contain multiple results), or truncate the file and start over (i.e. the file will always have a single partial or completed result).

The closest thing to a use case I can imagine is crm_mon's HTML output -- someone could run crm_mon continuously, and it could update a file that would be served by a web server. That would tend to the truncate approach. But the way to do this now is calling crm_mon via cron, and that's still a good approach. So again, we probably just don't need to care much at this stage.

If someone does follow the above flow with a file, an empty result will be printed first. The caller can avoid that by using this instead of reset:

    if (first_time) first_time = false;
    else reset(out);